### PR TITLE
[coqargs] use the standard option injection system for -w

### PR DESCRIPTION
**Pull request by gares**
Here the separate PR about the warning system, to allow for a cleaner review.

About the doc, @SkySkimmer I don't really think this is necessary.
The `current` keyword is used internally, since the current semantics is that `Set Warning...` "adds" to the current set of warnings.
If we decide to change that, or to provide another command which really sets all warnings, then `current` may become user visible (CC @Zimmi48) and require doc.

### CODE REVIEW

**Comment by gares** [comment]
Hum, maybe I did misunderstand the current code. `-w` is cumulative, but is `Set Warnings "...."` so?
If not this is new user facing "syntax" and needs to be documented.

**Comment by SkySkimmer** [comment]
Set Warnings is cumulative, see https://github.com/coq/coq/blob/0a6444c522c18c634fe1030436ea82f326bada9a/vernac/vernacentries.ml#L1708-L1711

**Comment by gares** [comment]
Do you want me to change that code to use `current` and remove the comment?

**Comment by SkySkimmer** [comment]
I don't see the benefit of this change tbh.

**Comment by gares** [comment]
This PR does 2 changes:
- using the injection to set options
- `current`

The former is what I need, but in order to use the injection one cannot grab the current status *at argument parsing time* as the old code is doing, since injections are fired later (at document creation time).
The `current` thing seemed a better way to implement the additive semantics in a non-imperative way, but one can hack things around in another way, if you really don't like it.

**Comment by gares** [comment]
Let me explain better, with the new code one puts in a pipeline `Set Warnings "current,+foo".` and then `Set Warnings "current,+bar". So it does not matter when these two are run, the latter will, thanks to current, add to the former.
If you expand `current` in these two ahead of time, then it is going to be empty, and only `+bar` will be set.

**Comment by SkySkimmer** [comment]
IIRC I intentionally had -w act immediately so that it would work on early warnings, for instance `coqc -w -deprecated -sprop-cumulative foo.v` wouldn't warn. 
I guess you have a use case for using the goptions system which requires giving up on that?
If so IMO we can just have ` type option_command = OptionSet of string option | OptionAppend of string | OptionUnset` and use OptionAppend for -w without changing the rest of the warning stuff, or maybe special case the handling of OptionSet for warnings like we do in vernacentries (this would also make `-set Warnings=bla` append, which is more consistent)

**Comment by gares** [comment]
Yes, as you can see in the other PR I tried to make "argument parsing" only *parse* argument and not set global variables on the spot, so that the initialization of the system can be controlled a bit more (I did not finish, I did the ones that broke badly).

IMO all the settings should be moved to goptions and injections, so their execution order can still be respected, but their execution time would not be dictated by the parser. Does this plan make sense to you?

About `OptionAppend` I did think a bit about that, but it seemed warnings are the exception and not the norm. So I opted for the implementation with `current`. I can change that if you see other use cases.

**Comment by gares** [comment]
Yes, as you can see in the other PR I tried to make "argument parsing" only *parse* argument and not set global variables on the spot, so that the initialization of the system can be controlled a bit more (I did not finish, I did the ones that broke badly).

IMO all the settings should be moved to goptions and injections, so their execution order can still be respected, but their execution time would not be dictated by the parser. Does this plan make sense to you?

About `OptionAppend` I did think a bit about that, but it seemed warnings are the exception and not the norm. So I opted for the implementation with `current`. I can change that if you see other use cases.

**Comment by Zimmi48** [comment]
I like this plan, personally. I didn't review the code (yet).

**Comment by gares** [comment]
Since this PR blocks #13418 I'd like to get feedback on it quickly, please.

I need to make command line argument parsing *pure*, but I don't have strong opinions on *how* this should be done. So if this PR needs to be amended, just tell me.

**Comment by SkySkimmer** [comment]
Please do OptionAppend

**Comment by gares** [comment]
done, I did not realize that there was already some support for appending.

File: toplevel/coqargs.ml [hunk 1]
```diff
@@ -464,12 +464,8 @@ let parse_args ~help ~init arglist : t * string list =
 
     |"-w" | "-W" ->
       let w = next () in
-      if w = "none" then
-        (CWarnings.set_flags w; oval)
-      else
-        let w = CWarnings.get_flags () ^ "," ^ w in
-        CWarnings.set_flags (CWarnings.normalize_flags_string w);
-        oval
+      if w = "none" then add_set_option oval ["Warnings"] (Stm.OptionSet(Some w))
```
Comment (SkySkimmer) [inline toplevel/coqargs.ml:467]: Why this special case?

File: library/goptions.ml [hunk 2]
```diff
@@ -422,9 +422,16 @@ let set_option_value ?(locality = OptDefault) check_and_cast key v =
   let opt = try Some (get_option key) with Not_found -> None in
   match opt with
   | None -> warn_unknown_option key
-  | Some (depr, (read,write,append)) ->
+  | Some (depr, (read,write,_)) ->
     write locality (check_and_cast v (read ()))
 
+let append_option_value ?(locality = OptDefault) check_and_cast key v =
```
Comment (SkySkimmer) [inline library/goptions.ml:428]: Doesn't this duplicate `set_string_option_append_value_gen`? (if you want to rename it that's fine, I don't even know why it's called _gen)

File: stm/stm.ml [hunk 3]
```diff
@@ -2404,7 +2407,8 @@ let new_doc { doc_type ; ml_load_path; vo_load_path; injections; stm_options } =
   let set_option = let open Goptions in function
       | opt, OptionUnset -> unset_option_value_gen ~locality:OptLocal opt
       | opt, OptionSet None -> set_bool_option_value_gen ~locality:OptLocal opt true
-      | opt, OptionSet (Some v) -> set_option_value ~locality:OptLocal (interp_set_option opt) opt v in
+      | opt, OptionSet (Some v) -> set_option_value ~locality:OptLocal (interp_set_option opt) opt v
+      | opt, OptionAppend v -> append_option_value ~locality:OptLocal (interp_set_option opt) opt v in
```
Comment (SkySkimmer) [inline stm/stm.ml:2411]: Append is only for string so we can do
```suggestion
      | opt, OptionAppend v -> set_string_option_append_value_gen ~locality:OptLocal opt v in
```
and don't need to define `append_option_value`

(not sure why it's called `_gen`, we could rename it)

File: toplevel/coqargs.ml (hunk 1 above)
Comment (gares) [reply toplevel/coqargs.ml:467]: It was already there. IIRC the code in cWarning checks if the *entire* string is "none" and in that case it resets.
Maybe I still don't get exactly how the goption thing works.

File: stm/stm.ml (hunk 3 above)
Comment (gares) [reply stm/stm.ml]: ok, done

File: toplevel/coqargs.ml (hunk 1 above)
Comment (SkySkimmer) [reply toplevel/coqargs.ml:467]: >It was already there

Right, I read too fast. It looks like premature optimisation but it's fine I guess.

File: library/goptions.ml [hunk 4]
```diff
@@ -422,7 +422,7 @@ let set_option_value ?(locality = OptDefault) check_and_cast key v =
   let opt = try Some (get_option key) with Not_found -> None in
   match opt with
   | None -> warn_unknown_option key
-  | Some (depr, (read,write,append)) ->
+  | Some (depr, (read,write,_)) ->
```
Comment (SkySkimmer) [inline library/goptions.ml:425]: spurious change

Comment (gares) [reply library/goptions.ml:425]: so what? it's better than before, no?

**Comment by gares** [comment]
done

File: stm/stm.ml [hunk 5]
```diff
@@ -2404,7 +2407,8 @@ let new_doc { doc_type ; ml_load_path; vo_load_path; injections; stm_options } =
   let set_option = let open Goptions in function
       | opt, OptionUnset -> unset_option_value_gen ~locality:OptLocal opt
       | opt, OptionSet None -> set_bool_option_value_gen ~locality:OptLocal opt true
-      | opt, OptionSet (Some v) -> set_option_value ~locality:OptLocal (interp_set_option opt) opt v in
+      | opt, OptionSet (Some v) -> set_option_value ~locality:OptLocal (interp_set_option opt) opt v
+      | opt, OptionAppend v -> set_string_option_value_gen ~locality:OptLocal opt v in
```
Comment (Zimmi48) [inline stm/stm.ml:2411]: I don't understand the new code, in particular how this `set_string_option_value_gen` is any different from the `set_option_value` above (given that it is precisely defined in terms of `set_option_value`).

File: toplevel/coqargs.ml [hunk 6]
```diff
@@ -464,12 +464,8 @@ let parse_args ~help ~init arglist : t * string list =
...
+      else add_set_option oval ["Warnings"] (Stm.OptionAppend w)
```
Comment (Zimmi48) [inline toplevel/coqargs.ml:468]: This special casing is definitely useless if the two calls end up doing the exact same thing (see my previous comment / question).

File: stm/stm.ml (hunk 5 above)
Comment (gares) [reply stm/stm.ml]: oops

**Comment by gares** [comment]
it should be ok now

**Review by SkySkimmer — APPROVED** [review]

**Comment by gares** [comment]
CI green

**Comment by SkySkimmer** [comment]
@coqbot: merge now

**Comment by gares** [comment]
thanks

**Pull request merged**